File: app/src/main/java/org/freelesson/pagingsample/ui/SearchRepositoriesActivity.java
package org.freelesson.pagingsample.ui;

import android.os.Bundle;
import android.os.PersistableBundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.freelesson.pagingsample.Injection;
import org.freelesson.pagingsample.R;


public class SearchRepositoriesActivity extends AppCompatActivity {
    private static final String LAST_SEARCH_QUERY = "last_search_query";
    private static final String DEFAULT_QUERY = "Android";
    SearchRepositoriesViewModel viewModel;
    ReposAdapter adapter = new ReposAdapter();
    RecyclerView list;
    TextView emptyList;
    EditText search_repo;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_repositories);
        emptyList = findViewById(R.id.emptyList);
        list = findViewById(R.id.list);
        search_repo = findViewById(R.id.search_repo);
        viewModel = new ViewModelProvider(this, Injection.provideViewModelFactory(this)).get(SearchRepositoriesViewModel.class);
        DividerItemDecoration decoration = new DividerItemDecoration(this,DividerItemDecoration.VERTICAL);

        list.addItemDecoration(decoration);
        setupScrollListener();

        initAdapter();
        String query = savedInstanceState!= null && savedInstanceState.getString(LAST_SEARCH_QUERY)!=null ? savedInstanceState.getString(LAST_SEARCH_QUERY) : DEFAULT_QUERY;
        viewModel.searchRepo(query);
        initSearch(query);

    }

    @Override
    public void onSaveInstanceState(Bundle outState, PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);
        outState.putString(LAST_SEARCH_QUERY,viewModel.lastQueryValue());
    }

    private void initSearch(String query) {
        search_repo.setText(query);
        search_repo.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId== EditorInfo.IME_ACTION_GO) {
                updateRepoListFromInput();
                return true;
            } else {
                return false;
            }
        });

        search_repo.setOnKeyListener((v, keyCode, event) -> {
            if (event.getAction() == KeyEvent.ACTION_DOWN && keyCode==KeyEvent.KEYCODE_ENTER) {
                updateRepoListFromInput();
                return true;
            }
            return false;
        });
    }
    private void updateRepoListFromInput() {
        if (!search_repo.getText().toString().trim().isEmpty()) {
            list.scrollToPosition(0);
            viewModel.searchRepo(search_repo.getText().toString().trim());
            adapter.submitList(null);
        }
    }
    private void initAdapter() {
        list.setAdapter(adapter);
        viewModel.repos.observe(this, repos -> {
            Log.d("Activity","list "+repos.size());
            showEmptyList(repos.isEmpty());
            adapter.submitList(repos);
        });
        viewModel.networkErrors.observe(this, s -> Toast.makeText(getApplicationContext(),"\uD83D\uDE28 Wooops "+s,Toast.LENGTH_LONG).show());

    }

    private void setupScrollListener() {
        final LinearLayoutManager layoutManager = (LinearLayoutManager) list.getLayoutManager();
        list.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView,dx,dy);
                assert layoutManager != null;
                int totalItemCount = layoutManager.getItemCount();
                int visibleItemCount = layoutManager.getChildCount();
                int lastVisibleItem = layoutManager.findLastVisibleItemPosition();
                viewModel.listScrolled(visibleItemCount,lastVisibleItem,totalItemCount);
            }
        });
    }
    private void showEmptyList(boolean show) {
        if (show) {
            emptyList.setVisibility(View.VISIBLE);
            list.setVisibility(View.GONE);
        } else {
            emptyList.setVisibility(View.GONE);
            list.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/org/freelesson/pagingsample/ui/SearchRepositoriesActivity.kt
package org.freelesson.pagingsample.ui

import android.os.Bundle
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.paging.LoadState
import androidx.paging.PagingData
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import org.freelesson.pagingsample.Injection
import org.freelesson.pagingsample.databinding.ActivitySearchRepositoriesBinding

class SearchRepositoriesActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivitySearchRepositoriesBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        val viewModel = ViewModelProvider(
            this,
            Injection.provideViewModelFactory(context = this, owner = this)
        )[SearchRepositoriesViewModel::class.java]
        val decoration = DividerItemDecoration(this, DividerItemDecoration.VERTICAL)
        binding.list.addItemDecoration(decoration)
        //bind the state
        binding.bindState(
            viewModel.state,
            pagingData = viewModel.pagingDataFlow,
            uiActions = viewModel.accept
        )
    }

    private fun ActivitySearchRepositoriesBinding.bindState(
        uiState: StateFlow<UiState>,
        pagingData: Flow<PagingData<UiModel>>,
        uiActions: (UiAction) -> Unit
    ) {
        val reposAdapter = ReposAdapter()
        val header = ReposLoadStateAdapter {
            reposAdapter.retry()
        }
        list.adapter = reposAdapter.withLoadStateHeaderAndFooter(
            header = header,
            footer = ReposLoadStateAdapter { reposAdapter.retry() })
        bindSearch(uiState = uiState, onQueryChanged = uiActions)
        bindList(
            header = header,
            reposAdapter = reposAdapter,
            uiState = uiState,
            pagingData = pagingData,
            onScrollChanged = uiActions
        )
    }

    private fun ActivitySearchRepositoriesBinding.bindSearch(
        uiState: StateFlow<UiState>,
        onQueryChanged: (UiAction.Search) -> Unit
    ) {
        searchRepo.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_GO) {
                updateRepoListFromInput(onQueryChanged)
                true
            } else {
                false
            }
        }
        searchRepo.setOnKeyListener { _, keyCode, event ->
            if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_ENTER) {
                updateRepoListFromInput(onQueryChanged)
                true
            } else {
                false
            }
        }
        lifecycleScope.launch {
            uiState.map { it.query }
                .distinctUntilChanged()
                .collect(searchRepo::setText)
        }


    }

    private fun ActivitySearchRepositoriesBinding.bindList(
        header: ReposLoadStateAdapter,
        reposAdapter: ReposAdapter,
        uiState: StateFlow<UiState>,
        pagingData: Flow<PagingData<UiModel>>,
        onScrollChanged: (UiAction.Scroll) -> Unit
    ) {

        retryButton.setOnClickListener { reposAdapter.retry() }
        list.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy != 0) onScrollChanged(UiAction.Scroll(currentQuery = uiState.value.query))
            }
        })
        val notLoading = reposAdapter.loadStateFlow
            .asRemotePresentationState()
            .map { it == RemotePresentationState.PRESENTED }
        val hasNotScrolledForCurrentSearch = uiState
            .map { it.hasNotScrolledForCurrentSearch }
            .distinctUntilChanged()
        val shouldScrollToTop =
            combine(notLoading, hasNotScrolledForCurrentSearch, Boolean::and).distinctUntilChanged()
        lifecycleScope.launch {
            pagingData.collectLatest(reposAdapter::submitData)
        }
        lifecycleScope.launch {
            shouldScrollToTop.collect { shouldScrollToTop ->
                if (shouldScrollToTop) list.scrollToPosition(0)
            }
        }
        lifecycleScope.launch {
            reposAdapter.loadStateFlow.collect { loadState ->
                header.loadState =
                    loadState.mediator?.refresh?.takeIf { it is LoadState.Error && reposAdapter.itemCount > 0 }
                        ?: loadState.prepend
                val isListEmpty =
                    loadState.refresh is LoadState.NotLoading && reposAdapter.itemCount == 0
                emptyList.isVisible = isListEmpty
                list.isVisible =
                    loadState.source.refresh is LoadState.NotLoading || loadState.mediator?.refresh is LoadState.NotLoading
                progressBar.isVisible = loadState.mediator?.refresh is LoadState.Loading
                retryButton.isVisible =
                    loadState.mediator?.refresh is LoadState.Error && reposAdapter.itemCount == 0
                val errorState = loadState.source.append as? LoadState.Error
                    ?: loadState.source.prepend as? LoadState.Error
                    ?: loadState.append as? LoadState.Error ?: loadState.prepend as? LoadState.Error
                errorState?.let {
                    Toast.makeText(
                        this@SearchRepositoriesActivity,
                        "\uD83D\uDE28 Wooops ${it.error}",
                        Toast.LENGTH_LONG
                    ).show()
                }
            }
        }
    }

    private fun ActivitySearchRepositoriesBinding.updateRepoListFromInput(onQueryChanged: (UiAction.Search) -> Unit) {
        searchRepo.text.trim().let {
            if (it.isNotEmpty()) {
                list.scrollToPosition(0)
                onQueryChanged(UiAction.Search(query = it.toString()))
            }
        }
    }

}